projects/plugins/jetpack/_inc/client/traffic/seo/custom-seo-titles.jsx

```diff
@@ -0,0 +1,279 @@
+		date: __( 'Date', 'jetpack' ),
+	},
+	tokensAvailablePerPageType: {
+		front_page: [ 'site_name', 'tagline' ],
+		posts: [ 'site_name', 'tagline', 'post_title' ],
+		pages: [ 'site_name', 'tagline', 'page_title' ],
+		groups: [ 'site_name', 'tagline', 'group_title' ],
+		archives: [ 'site_name', 'tagline', 'date' ],
+	},
+};
+
+/**
+ * Converts an input value to an array of token/string objects,
+ * for storage into the `advanced_seo_title_formats` option.
+ *
+ * @param {string} inputValue - The value of an input for one of the custom SEO title inputs/pageTypes.
+ * @param {string} pageType - Type of page the title is being customized for (e.g front_page, archives)
+ * @returns {Array} An array of token/string objects and their values.
+ */
+export const stringToTokenizedArray = ( inputValue, pageType ) => {
+	const inputArray = inputValue.split(
+		/(\[(?:site_name|tagline|post_title|page_title|group_title|date)\])/
+	);
+
+	return inputArray
+		.filter( value => {
+			if ( value ) {
+				return value;
+			}
+		} )
+		.map( value => {
+			let matchedToken = null;
+			Object.keys( customSeoTitleFormats.insertableTokens ).map( token => {
+				if ( value === `[${ token }]` ) {
+					matchedToken = token;
+				}
+			} );
+
+			if (
+				matchedToken &&
+				customSeoTitleFormats.tokensAvailablePerPageType[ pageType ].includes( matchedToken )
+			) {
+				return {
+					type: 'token',
+					value: matchedToken,
+				};
+			}
+
+			return {
+				type: 'string',
+				value,
+			};
+		} );
+};
+
+/**
+ * Converts an array of token/string objects into a single string.
+ * Objects with type of 'token' have their values enclosed by '[]'
+ *
+ * @param {Array} arr - An array of token/string objects and their values.
+ * @returns {string} A concatenated string of values from the token/string objects supplied.
+ */
+export const tokenizedArrayToString = arr => {
+	if ( Array.isArray( arr ) ) {
+		return arr.reduce( ( acc, obj ) => {
+			return acc + ( obj.type === 'token' ? `[${ obj.value }]` : obj.value );
+		}, '' );
+	}
+
+	return '';
+};
+
+const getCustomSeoTitleInputPreview = ( pageType, value, siteData ) => {
+	customSeoTitleFormats.tokensAvailablePerPageType[ pageType.name ].forEach( token => {
+		switch ( token ) {
+			case 'site_name':
+				value = value.replace( /\[site_name\]/g, siteData.title );
+				break;
+			case 'tagline':
+				value = value.replace( /\[tagline\]/g, siteData.tagline );
+				break;
+			case 'post_title':
+				value = value.replace( /\[post_title\]/g, __( 'Example Title', 'jetpack' ) );
+				break;
+			case 'page_title':
+				value = value.replace( /\[page_title\]/g, __( 'Example Title', 'jetpack' ) );
+				break;
+			case 'group_title':
+				value = value.replace( /\[group_title\]/g, __( 'Tag', 'jetpack' ) );
+				break;
+			case 'date':
+				value = value.replace( /\[date\]/g, dateI18n( 'F Y', Date.now() ) );
+				break;
+			default:
+				break;
+		}
+	} );
+
+	return __( 'Preview: ', 'jetpack' ) + value;
```

Comment:
I think it may be useful to have this being one string with `value` as a placeholder.

Reply:
Agreed, https://github.com/Automattic/jetpack/pull/18810/commits/abfc9a00b7e4c7047401a26b42537f782fe3ae79